sdk/typescript/src/rpc/client.ts

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import RpcClient from 'jayson/lib/client/browser';
+import {
+  literal,
+  type as pick,
+  string,
+  Struct,
+  unknown,
+  assert,
+  optional,
+  any,
+  is,
+} from 'superstruct';
+import fetch from 'cross-fetch';
+
+/**
+ * An object defining headers to be passed to the RPC server
+ */
+export type HttpHeaders = { [header: string]: string };
+
+export class JsonRpcClient {
+  private rpcClient: RpcClient;
+
+  constructor(url: string, httpHeaders?: HttpHeaders) {
+    this.rpcClient = this.createRpcClient(url, httpHeaders);
+  }
+
+  private createRpcClient(url: string, httpHeaders?: HttpHeaders): RpcClient {
+    const client = new RpcClient(async (request, callback) => {
+      const options = {
+        method: 'POST',
+        body: request,
+        headers: Object.assign(
+          {
+            'Content-Type': 'application/json',
+          },
+          httpHeaders || {}
+        ),
+      };
+
+      try {
+        let res: Response = await fetch(url, options);
+        const text = await res.text();
+        if (res.ok) {
+          callback(null, text);
+        } else {
+          callback(new Error(`${res.status} ${res.statusText}: ${text}`));
+        }
+      } catch (err) {
+        if (err instanceof Error) callback(err);
+      }
+    }, {});
+
+    return client;
+  }
+
+  async requestWithType<T, S>(
```

Comment:
we used to have nice methods per route, which seems preferable to a a method where you pass a string.  could we return to that model ?

Reply:
We do have a nice method per route defined in 

https://github.com/MystenLabs/sui/blob/b671e70ece84dbd53bccedd56c5b2941ccb0d523/sdk/typescript/src/providers/json-rpc-provider.ts#L27

End users will not use this client directly

Reply:
> > No other web3 sdk ships with a redundant data validation layer.
> 
> That's not true. Solana uses it [extensively](https://github.com/solana-labs/solana-web3.js/blob/4d4a0faa8b9a6671bb5e94e8548fba746c3e4abc/src/connection.ts#L21).
> 
> > It adds a lot of code, but we don't seem to get much benefit from it.
> 
> I don't think it adds much code since we need those type definitions anyway and the package itself is very minimal. The benefit we get is to have a clear error boundary so that developers can safely assume that the data they receive is correct and we can write integration tests to always making sure the SDK definition matches what're returned from the gateway. This is important as we are making rapid iterations before mainnet launch.
> 
> > we could use Typescript type guards
> 
> Those are much verbose to define and validate than `superstruct`
> 
> Given that this library is used in a major web3 sdk and there's no major downside of using it, I would much preferred to keep it instead of rewriting the module since I have other higher priorities to work on. Can we revisit the decision after devnet?

It still seems like fundamentally, if the data source which we control is returning data which isn't formatted correctly, this is the real problem which should be fixed - what is the user to do if Superstruct invalidates the data ?

I also feel that using superstruct provides for less readable code & less of the benefit of using typescript - it's "designed for typescript" but looking at the superstruct definitions feels like somebody trying to emulate typescript at runtime.

superstruct:
```typescript
export const ObjectRef = pick({
  digest: string(),
  objectId: string(),
  version: number(),
});
```

normal typescript:
```typescript
export type ObjectRef = {
  digest: string,
  objectId: string,
  version: number,
};
```

I think that it seems unlikely we would come back to replace it if we ship it now.

Reply:
> It still seems like fundamentally, if the data source which we control is returning data which isn't formatted correctly, this is the real problem which should be fixed - what is the user to do if Superstruct invalidates the data ?

A good mental model for the system is that they are built in the following layers:

```
1. Application Code
-----------------
2. SDK
----------------
3. Gateway
-----------------
.........
```

In general, we should define clear error boundary and interface between each layers. For example,

```
async getOwnedObjectRefs(address: string): Promise<ObjectRef[]>
```
is the interface/contract between Application code and the SDK. When the SDK says it will return `ObjectRef[]`, the application code should feel safe to assume that this is the case, instead of needing to check whether the data returned is indeed `ObjectRef[]`, which is why runtime validation and returning an error is important within the SDK.

If we don't do runtime validation within the SDK, application code has to do it anyway in their code, which is not only unprincipled, but also put more burden on the application developers.

> what is the user to do if Superstruct invalidates the data ?

They will receive an exception and knows that our APIs are broken(and can submit a PR to fix) instead of trying to debug if it's their problem or the SDK problem.

>  it's "designed for typescript" but looking at the superstruct definitions feels like somebody trying to emulate typescript at runtime.

`superstruct` solves the validation problem instead of just emulating typescript. It's super well-received in the TS community from the [articles](https://www.google.com/search?q=why+use+superstruct&oq=why+u&aqs=chrome.0.69i59l3j69i57j0i433i512j69i60l3.1168j0j7&sourceid=chrome&ie=UTF-8) I've read